### PR TITLE
feat(vehicle): Vehicle Ability Selector

### PR DESCRIPTION
**Pull request by TheGreenBandit**
i have not tested all cars for their abilities. Also I have tested deluxo and it did not return a value, so this pr wont solve that issue.

(This pr cannot be merged until the pr in gtavclasses is merged)
(also i have not bothered to put it in the translations yet)

to find flags input in the overlay the following:
```
			if (g_local_player && g_local_player->m_vehicle)
			{
				CVehicleModelInfo* modelinfo = (CVehicleModelInfo*)g_local_player->m_vehicle->m_model_info;
				ImGui::Text(std::format("Vehicle Flag {}", modelinfo->m_ability_flag).c_str());
			}
```

Footage of ingame

https://www.youtube.com/watch?v=3AbLWN3woEU

### CODE REVIEW

**Comment by 0xEmiil** [comment]
Wao, Not gonna lie, the vehicle parachute its really nice!

**Comment by TheGreenBandit** [comment]
This will confict with vehicle jump as shown in the video.

File: src/backend/looped/vehicle/vehicle_abilities.cpp [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+				PAD::DISABLE_CONTROL_ACTION(0, (int)ControllerInputs::INPUT_VEH_EXIT, 0);
+
+				if (PED::IS_PED_DEAD_OR_DYING(self::ped, 0) || PAD::IS_DISABLED_CONTROL_JUST_RELEASED(0, (int)ControllerInputs::INPUT_VEH_EXIT))
+				{
+					g.vehicle.ability_chosen  = VehicleAbility::NONE;
+					modelinfo->m_ability_flag = 0;
+					TASK::TASK_LEAVE_VEHICLE(self::ped, self::veh, 0);
+				}
+
+				if (g.vehicle.ability_chosen == VehicleAbility::BOOST)
```
Comment (Tayipc38) [inline src/backend/looped/vehicle/vehicle_abilities.cpp]: A switch case looks more suitable here

File: src/backend/looped/vehicle/vehicle_abilities.cpp [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+namespace big
+{
+	class ability_selector : looped_command
+	{
+		using looped_command::looped_command;
+
+		virtual void on_enable() override
+		{
+			if (self::veh)
+				PAD::DISABLE_CONTROL_ACTION(0, (int)ControllerInputs::INPUT_VEH_EXIT, 0);
```
Comment (Tayipc38) [inline src/backend/looped/vehicle/vehicle_abilities.cpp]: This native only runs a single frame

File: src/backend/looped/vehicle/vehicle_abilities.cpp (hunk 1 above)
Comment (TheGreenBandit) [reply src/backend/looped/vehicle/vehicle_abilities.cpp:31]: something like 	switch (g.vehicle.ability_chosen)
				{
				case VehicleAbility::BOOST: modelinfo->m_ability_flag = 0; break;
				}

Comment (Tayipc38) [reply src/backend/looped/vehicle/vehicle_abilities.cpp]: Yes

File: src/views/vehicle/view_vehicle.cpp [hunk 3]
```diff
@@ -177,6 +178,18 @@ namespace big
 
 				ImGui::EndCombo();
 			}
+			CVehicleModelInfo* modelinfo = (CVehicleModelInfo*)g_local_player->m_vehicle->m_model_info;
+
+			ImGui::InputInt("Ability Flag Value", (int*)g.vehicle.customvalue, 1, 10);
```
Comment (maybegreat48) [inline src/views/vehicle/view_vehicle.cpp]: `&g.vehicle.customvalue`

**Comment by deepdream88** [comment]
Very cool PR, hope it gets merged soon,Thanks for your coding

File: src/views/vehicle/view_vehicle.cpp (hunk 3 above)
Comment (TheGreenBandit) [reply src/views/vehicle/view_vehicle.cpp:183]: (int*) & g.vehicle.customvalue?

Comment (Yimura) [reply src/views/vehicle/view_vehicle.cpp]: Yes

**Comment by xiaoxiao921** [comment]
i'm personally against merging this PR until you properly define the enum flag and make the GUI a checkbox based UI
an enum flag usually look like this:
```
enum Flags
{
    A = 1 << 0, // binary 0001
    B = 1 << 1, // binary 0010
    C = 1 << 2, // binary 0100
    D = 1 << 3  // binary 1000
};
```

when you write things like `case VehicleAbility::BOOSTPARACHUTERAMPBUGGY: modelinfo->m_ability_flag = 832; break;` just realize that what you are writing is `0b1101000000` which is just 3 bit being set to 1

setting a bit flag
modelinfo->m_ability_flag |= VehicleAbility::RampBuggy;

removing a bit flag
modelinfo->m_ability_flag = modelinfo->m_ability_flag & (~VehicleAbility::RampBuggy);

**Comment by navmodder** [comment]
> i'm personally against merging this PR until you properly define the enum flag and make the GUI a checkbox based UI an enum flag usually look like this:
> 
> ```
> enum Flags
> {
>     A = 1 << 0, // binary 0001
>     B = 1 << 1, // binary 0010
>     C = 1 << 2, // binary 0100
>     D = 1 << 3  // binary 1000
> };
> ```
> 
> when you write things like `case VehicleAbility::BOOSTPARACHUTERAMPBUGGY: modelinfo->m_ability_flag = 832; break;` just realize that what you are writing is `0b1101000000` which is just 3 bit being set to 1
> 
> setting a bit flag modelinfo->m_ability_flag |= VehicleAbility::RampBuggy;
> 
> removing a bit flag modelinfo->m_ability_flag = modelinfo->m_ability_flag & (~VehicleAbility::RampBuggy);
why make it complicate just use ints they will do same job

 example:
rocket boost + parachute = 40 01 00 00 00 00 00 00 00 00
vehicle jump + parachute = 20 01 00 00 00 00 00 00 00 00
rocket boost + jump = E7 00 00 00 00 00 00 00 00 00
others whatever 

i mean if you have free time. just go and make log each bits and tell us. + instead off looping this you can make it just buttons to set the value. and you can also get vehicle flags before entering into submenu and check if they exist and change button color or tick it like this ✓. will be much better. just a suggestion.

**Comment by xiaoxiao921** [comment]
No idea what you are saying, the field type is an enum flag, i'm just asking you to define it as is, simple as that, make it easier for everyone and allow mix and match however you want

**Comment by navmodder** [comment]
sorry englitch is bad

**Comment by xiaoxiao921** [comment]
Please check `allowed_entity_types` in the YimMenu code base,

from using `misc::has_bits_set` to the usage of `ImGui::CheckboxFlags`

**Comment by TheGreenBandit** [comment]
> Iam trying to impliment this into my version of the menu but keep getting m_ability_flag is not a member of cvehiclemodelinfo  any help?

You have to change the classes manually adding it.

**Comment by Yimura** [comment]
Stale PR, closing it for now.

Related to the GTAV-Classes that has been closed for the same reason.

**Pull request closed without merging**